spirv-builder/src/test/control_flow.rs

```diff
@@ -0,0 +1,162 @@
+use super::val;
+
+#[test]
+fn cf_while() {
+    val(r#"
+#[allow(unused_attributes)]
+#[spirv(fragment)]
+pub fn main(i: Input<i32>) {
+    while i.load() < 10 {
+    }
+}
+"#);
+}
+
+#[test]
+fn cf_while_while() {
+    val(r#"
+#[allow(unused_attributes)]
+#[spirv(fragment)]
+pub fn main(i: Input<i32>) {
+    while i.load() < 20 {
+        while i.load() < 10 {
+        }
+    }
+}
+"#);
+}
+
+#[test]
+fn cf_while_while_break() {
+    val(r#"
+#[allow(unused_attributes)]
+#[spirv(fragment)]
+pub fn main(i: Input<i32>) {
+    while i.load() < 20 {
+        while i.load() < 10 {
+            break;
+        }
+    }
+}
+"#);
+}
+
+#[test]
+fn cf_while_while_if_break() {
+    val(r#"
+#[allow(unused_attributes)]
+#[spirv(fragment)]
+pub fn main(i: Input<i32>) {
+    while i.load() < 20 {
+        while i.load() < 10 {
+            if i.load() > 10 {
+                break;
+            }
+        }
+    }
+}
+"#);
+}
+
+#[test]
+fn cf_while_break() {
+    val(r#"
+#[allow(unused_attributes)]
+#[spirv(fragment)]
+pub fn main(i: Input<i32>) {
+    while i.load() < 10 {
+        break;
+    }
+}
+"#);
+}
+
+#[test]
+fn cf_while_if_break() {
+    val(r#"
+#[allow(unused_attributes)]
+#[spirv(fragment)]
+pub fn main(i: Input<i32>) {
+    while i.load() < 10 {
+        if i.load() == 0 {
+            break;
+        }
+    }
+}
+"#);
+}
+
+#[test]
+fn cf_if() {
+    val(r#"
+#[allow(unused_attributes)]
+#[spirv(fragment)]
+pub fn main(i: Input<i32>) {
+    if i.load() > 0 {
+
+    }
+}
+"#);
+}
+#[test]
+fn cf_ifx2() {
+    val(r#"
+#[allow(unused_attributes)]
+#[spirv(fragment)]
+pub fn main(i: Input<i32>) {
+    if i.load() > 0 {
+
+    }
+    if i.load() > 1 {
+
+    }
+}
+"#);
+}
+
+#[test]
+fn cf_if_else() {
+    val(r#"
+#[allow(unused_attributes)]
+#[spirv(fragment)]
+pub fn main(i: Input<i32>) {
+    if i.load() > 0 {
+
+    } else {
+
+    }
+}
+"#);
+}
+
+#[test]
+fn cf_if_elseif_else() {
+    val(r#"
+#[allow(unused_attributes)]
+#[spirv(fragment)]
+pub fn main(i: Input<i32>) {
+    if i.load() > 0 {
+
+    } else if i.load() < 0 {
+
+    } else {
+        
+    }
```

Comment:
Rustc may optimize this control flow away. Maybe `-Zmir-opt-level=0` should be used?

Reply:
Discussed this with Ashley already :) 

`note they have the #[spirv(fragment)] attribute, so they're entrypoints and held on to by DCE`

`the function isn't inlined (because it only has plain arguments), which means that when considering the call, the call could do side effects or whatever (kinda? not on the gpu, but whatever), and it has no knowledge of the internals, so it pessimistically keeps it`

was her response.